crates/ide-completion/src/render/function.rs

```diff
@@ -153,6 +154,34 @@ fn render(
     item
 }
 
+/// When typing `::` of a type, the preferred order is:
+/// * Constructors: new like functions to be able to create the type,
+/// * Builder Methods,
+/// * Constructors that take args: Any other function that creates Self
+/// * Regular methods & Associated functions
+///
+fn calculate_bonus(ctx: &RenderContext<'_>, func: hir::Function, db: &dyn HirDatabase) -> u32 {
+    if ctx.token().kind() != syntax::SyntaxKind::COLON2 || func.self_param(db).is_some() {
```

Comment:
This should check whether we are completing an associated function or not instead of checking the token kind, `func.as_assoc_item(ctx.db()).is_some()` can be used for that